oqsprov/oqsprov_keys.c

```diff
@@ -87,13 +155,21 @@ void oqsx_key_free(OQSX_KEY *key)
 #endif
     if (refcnt > 0)
         return;
+#ifndef NDEBUG
     assert(refcnt == 0);
+#endif
 
     OPENSSL_free(key->propq);
     OPENSSL_secure_clear_free(key->privkey, key->privkeylen);
     OPENSSL_secure_clear_free(key->pubkey, key->pubkeylen);
-    if (key->iskem) OQS_KEM_free(key->key.k);
-    else OQS_SIG_free(key->key.s);
+    if (key->keytype == KEY_TYPE_KEM)
+        OQS_KEM_free(key->primitive.kem);
+    else if (key->keytype == KEY_TYPE_ECP_HYB_KEM || key->keytype == KEY_TYPE_ECX_HYB_KEM) {
```

Comment:
For my understanding: Are these two different key types really necessary? Wouldn't one suffice? Or would a third be needed when adding RSA3k?

Reply:
Technically one would suffice. Both `KEY_TYPE_ECP_HYB_KEM` and `KEY_TYPE_ECX_HYB_KEM` refer to the same code.

The hybrid implementation choses the classical scheme (ecp/ecx) implicitly using as input (1) a variable that determines which family to chose (ecp/ecx) and (2) the NIST security level of the pqc KEM. That way code is a bit more compact and avoids having to iterate through each individual combination. So the two different hybrid types are because of (1).

A third type would be necessary when combining with RSA.

Reply:
OK, understood, thanks. I also like more compact code in general, but wouldn't it be desirable to also support RSA to get to the same level of functionality with the provider as with the OSSL111 fork? If so, maybe another, more explicit approach might be better.

Reply:
Yeah RSA would be desirable, I will add that. Are there other families we should consider (e.g. Brainpool ECC, DH)?
If we have a relatively concluding list of families to support, we can for each of them just define another `KEY_TYPE_` (at the end its just an enum) and code changes are minimal.
Is your concern related to yml-ability of these combinations?

Reply:
My concern (goal, really) is the possibility to do interop testing with OQS-OSSL111 (at https://test.openquantumsafe.org) with the provider. So, Yes, there is a very concluding list: https://github.com/open-quantum-safe/openssl/blob/OQS-OpenSSL_1_1_1-stable/oqs-template/generate.yml (unless someone adds to it :-).

For the KEMs, this PR already supports more than what OQS-OSSL111 does (which is "only" strength-matching P256/P384/P521) if I get it right; for SIGs, RSA3k would be required to get to the point of being able to do interop testing (as and when we get around to implement serialization and later, OSSL3 gets provider-signature support: So, RSA3k is definitely not urgent right now but I thought it might be easy for you to add while you're "at it").

Reply:
Ok, thanks. I think I misinterpreted RSA3k as RSA-KEM (as in the sample OSSL3 provider).

Yes, this PR does than OQS-OSSL111 in the sense that it also does strength-matching with X25519 and X448 in addition to P256/P384/P521. I'd add another field (e.g. `nid_hybrid_ecx`) to generate.yml to reflect that.

For hybrid signatures, it seems that we'd need ECDSA P256/P384/P521 and RSA3k to do full interop testing. A more explicit approach for the keys could make sense there, will keep that in mind. I might first conclude this PR though.